core/chains/evm/client/tx_simulator.go

```diff
@@ -0,0 +1,117 @@
+package client
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink/v2/common/config"
+
+	commonclient "github.com/smartcontractkit/chainlink/v2/common/client"
+)
+
+const ErrOutOfCounters = "not enough keccak counters to continue the execution"
+
+type simulatorClient interface {
+	CallContext(ctx context.Context, result interface{}, method string, args ...interface{}) error
+}
+
+// ZK Chain can return an overflow error based on the number of keccak hashes in the call
+// This method allows a caller to determine if a tx would fail due to overflow error by simulating the transaction
+// Used as an entry point for custom simulation across different chains
+func SimulateTransaction(ctx context.Context, client simulatorClient, lggr logger.SugaredLogger, chainType config.ChainType, msg ethereum.CallMsg) error {
+	var err error
+	switch chainType {
+	case config.ChainZkEvm:
+		err = simulateTransactionZkEvm(ctx, client, lggr, msg)
+	default:
+		err = simulateTransactionDefault(ctx, client, msg)
+	}
+	// ClassifySendError will not have the proper fields for logging within the method due to the empty Transaction passed
+	code := ClassifySendError(err, lggr, &types.Transaction{}, msg.From, chainType.IsL2())
+	if code == commonclient.OutOfCounters {
+		return errors.New(ErrOutOfCounters)
+	}
+	return nil
+}
+
+// eth_estimateGas returns out-of-counters (OOC) error if the transaction would result in an overflow
+func simulateTransactionDefault(ctx context.Context, client simulatorClient, msg ethereum.CallMsg) error {
+	var result hexutil.Big
+	errCall := client.CallContext(ctx, &result, "eth_estimateGas", toCallArg(msg), "pending")
+	jsonErr, _ := ExtractRPCError(errCall)
+	// Only return error if Zk OOC error is identified
+	if jsonErr != nil && jsonErr.Message == ErrOutOfCounters {
+		return errors.New(ErrOutOfCounters)
+	}
+	return nil
+}
+
+// zkEVM implemented a custom zkevm_estimateCounters method to detect if a transaction would result in an out-of-counters (OOC) error
+func simulateTransactionZkEvm(ctx context.Context, client simulatorClient, lggr logger.SugaredLogger, msg ethereum.CallMsg) error {
+	var result struct {
+		CountersUsed struct {
+			GasUsed              string
+			UsedKeccakHashes     string
+			UsedPoseidonHashes   string
+			UsedPoseidonPaddings string
+			UsedMemAligns        string
+			UsedArithmetics      string
+			UsedBinaries         string
+			UsedSteps            string
+			UsedSHA256Hashes     string
+		}
```

Comment:
We could also check if the `CountersUsed` exceed the `CountersLimit` in the response? This way wouldn't be relying on the error message, which might be preferrable?

Reply:
Good point! I've added some validation checking these fields against each other. Since we're targeting OOC errors only though, I excluded the validation on `GasUsed`.